[#31] Replaced copy-pasted code with react-select library

FilipKopecky opened the pull request:
Removes copy-pasted code and replaces it with the library instead. The functionality should be the same as it was previously.
For better performance the react-window package was added. 

Resolves: lecbyjak/intelligent-tree-select#31

FilipKopecky commented:
I noticed that the selected value when multi is false is not clickable, gonna fix asap

ledsoft commented on src/components/IntelligentTreeSelect.js:
```diff
@@ -519,11 +521,12 @@ IntelligentTreeSelect.defaultProps = {
   options: [],
   renderAsTree: true,
-  isMenuOpen: false,
+  isMenuOpen: undefined,
```
Why change the default from `false` to `undefined`?

FilipKopecky replied:
You are right, it is not necessary to change it.

ledsoft commented on src/components/VirtualizedTreeSelect.js:
```diff
@@ -103,15 +109,36 @@ class VirtualizedTreeSelect extends Component {
+  _filterValues(searchInput) {
+    const matches = []
+    for (let option of this.state.options) {
+      if (option[this.props.labelKey].toLowerCase().indexOf(searchInput) !== -1) {
```
Should lowercase `searchInput` as well, otherwise this does not work (tested with demo). Also, I suggest extracting the match checking into a separate function, as we might want to extend the component API with support for custom matching functions (providing e.g. fuzzy string matching).

FilipKopecky replied:
I completely missed that, Thank you! The match checking in a separate function is a great idea, gonna implement that for sure 👨‍💻

ledsoft commented on src/components/VirtualizedTreeSelect.js:
```diff
@@ -180,11 +241,12 @@ VirtualizedTreeSelect.defaultProps = {
   optionLeftOffset: 16,
   expanded: false,
-  isMenuOpen: false,
+  isMenuOpen: undefined,
```
Same as in `IntelligentTreeSelect` - why change it from `false` to `undefined`. The code does not seem to require it.

ledsoft reviewed (changes requested):
Looks really good! But I ran into one issue - if I expand the tree deeper (e.g. 2-3 levels depth) and then try to collapse it directly at upper level, it does not hide the deeper descendants. E.g., expand `Equipment` (in the demo), then expand `Aircraft Performance`, then `Degraded performance` and then collapse `Aircraft performance` or `Equipment`, it still shows items under `Degraded performance`. If I collapse the tree gradually from the deepest expanded level, it has no problem. I assume it is just a problem of recursively collapsing descendants of the node being collapsed and should be easy to fix.
Otherwise see some minor comments inline.

FilipKopecky commented:
I implemented the proposed changes. When collapsing parent term all of the subterms of that particular term are collapsed as well, therefore hidden from the list. I added the possibility to add custom matchCheck(searchInput, inputLabel) function which filters out the results. 

While working on those changes I notices couple of bugs. First of them was the ability to expand/collapse items while using search. This behaviour is not allowed in the previous version so I disabled it here as well. Next mistake was the missing 'No results' component when search filtered out all of the results.

Please @ledsoft check it out and tell me what you think

ledsoft reviewed (changes requested):
I noticed one minor and one major issue when testing the component:
1. Typing into the input field and selecting an option makes the whole tree expand. It is not a problem per se, I am just curious if this is the intended behavior.
2. This is a bigger issue - the onScroll handler (passed as part of `listProps` to `VirtualizedTreeSelect`) is not invoked. It is essential for loading additional options, so it needs to be invoked. I checked the documentation of react-select and they do support `onScroll`, so it might be just a matter of updating the props and aligning the parameters of `onScroll` in `IntelligentTreeSelect` with the API of react-select.

FilipKopecky commented:
> 1. Typing into the input field and selecting an option makes the whole tree expand. It is not a problem per se, I am just curious if this is the intended behavior.

This was bugging me a lot, but in the previous version it is present as well. Once an item is expanded from the search, it doesn't automatically close, when the search is finished. I don't think this is the best UX. The expanded items from the search should automatically collapse when the searching is finished. Only the items I specifically toggled, should be opened. What do you think? @ledsoft 

> 2. This is a bigger issue - the onScroll handler (passed as part of `listProps` to `VirtualizedTreeSelect`) is not invoked. It is essential for loading additional options, so it needs to be invoked. I checked the documentation of react-select and they do support `onScroll`, so it might be just a matter of updating the props and aligning the parameters of `onScroll` in `IntelligentTreeSelect` with the API of react-select.

Regarding this, I am sorry that I didn't notice it. Will look into it

ledsoft commented:
> > 1. Typing into the input field and selecting an option makes the whole tree expand. It is not a problem per se, I am just curious if this is the intended behavior.
> 
> This was bugging me a lot, but in the previous version it is present as well. Once an item is expanded from the search, it doesn't automatically close, when the search is finished. I don't think this is the best UX. The expanded items from the search should automatically collapse when the searching is finished. Only the items I specifically toggled, should be opened. What do you think? @ledsoft

Good idea, I agree. I never noticed it in the previous version. However, I would also keep open the tree (branch) containing the selected item.

FilipKopecky commented:
Implemented requested changes. 
The `onScroll` function is now being passed properly to the component. It wasn't as straight forward as I thought it would be. I had to create a wrapper and use it's internal properties to pass the `onScroll` func. This is due to the fact that we are using the `react-window` for rendering and it's `onScroll` API  has completely different properties than the ones we need.

The search was modified as well. Now when user enters some text into it, the tree expands accordingly, but when the input is cleared only the items that were previously toggled by the user are expanded. Without this, the tree would be expanded from the search and not only by the user.

@ledsoft You mentioned that it would be nice to have the whole path to the selected item expanded. I don't know exactly what you mean by that. Well I have some questions regarding its intended behaviour.

1. When I select an item, the whole path to that item becomes **uncollapsable** until I deselect the value? Personally I don't like that but if thats something you would like to have, I will do it.
2. When user is searching for term via the text input and selects one of the terms, the path in the original tree should be expanded. That's the basic functionality right?

Thanks for answering the questions and sorry for the delay, I got overwhelmed by some other work last week.

ledsoft commented:
Filip,
thanks, nice work with the scrolling.
As for your questions, no. 2 is exactly what I meant - if a type something and select an item in the tree based on the displayed results, I would like the item's ancestor tree to remain expanded when the rest of the tree collapses after the input looses focus. However, I should still be able to collapse the tree containing selected items manually (that should answer question 1).
So, again:
1. No, that is definitely not the desired behavior.
2. Yes, exactly that is how I would expect it to work.

FilipKopecky commented:
Added the path expansion when item is selected from the tree while using the search feature.
The component now remembers the toggled items + items leading to selected items from the search, these are taken as they were toggled by the user. 
Please @ledsoft take a look at it, they are many states in which the tree can be in. I hope I tested every possible scenario, but there might be something I missed.

ledsoft reviewed (approved):
Seems to work nicely. Good job! I'll begin testing with TermIt.
